Add Roon volume hooks

pavoni opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Roon via their API provide a means of hooking the volume control in their app suite  and providing your own implementation of the volume control.

This can be useful in a situation where Home Assistant has control over the volume of a media player - but Roon natively does not.

Roon media players are often streamers connected to other devices. 

An  example would be an amplifier or similar device connected to a Roon streamer - where the amplifier has its own HA integration with control over the volume, or can respond to HA via an IR blaster.

This PR allows you to set up a volume control in the Roon apps. When this is changed - a device automation is fired in HA - allowing you to control the volume using HA's facilities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  TBA

(I need to update the HA roon documentation before this is merged, but otherwise I think this is ready for review)

This screenshot shows how the volume controls are exposed in the roon app:
<img width="704" alt="Screenshot 2023-09-06 at 10 25 12" src="https://github.com/home-assistant/core/assets/4487313/3e5eeb00-3ff2-47b1-b517-4fc663af0f5b">

And this shows the volume control buttons in the roon app:
<img width="514" alt="Screenshot 2023-09-06 at 10 18 27" src="https://github.com/home-assistant/core/assets/4487313/ae06e179-e47e-451a-bf9f-6a5be067bb6a">

This is the device trigger exposed in HA:
<img width="654" alt="Screenshot 2023-09-06 at 10 21 01" src="https://github.com/home-assistant/core/assets/4487313/80588f4e-ec29-497c-8566-4da0653ebf27">


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

pavoni committed:
First draft of device trigger.

pavoni committed:
Add config option.

pavoni committed:
Revise option handling - wip

pavoni committed:
Tidy option change handling.

pavoni committed:
Tidy.

pavoni committed:
Add tests.

pavoni committed:
Add config option test.

pavoni committed:
Remote trace.

pavoni committed:
Remove config listener.

pavoni committed:
Tidy.

pavoni committed:
Remove trace.

pavoni commented:
@MartinHjelmare you've been kind enough to review a number of the roon PR's before. 

Sure you are very busy - but would be very grateful for your views if you have time.

MartinHjelmare reviewed (commented):
Can't the user already create an automation that triggers on the roon media player entity volume state change?

pavoni commented:
@MartinHjelmare Good question. 

Roon allows you to choose a volume control method for each endpoint.

If Roon can control the volume itself directly (usually on devices that have fully integrated with Roon - which they call _Roon Ready_), then you can set _Device Volume_ - in the case the endpoint has a volume contro. This is the best options for devices that supported it - and this PR is unneeded. Sometimes (not on Roon Ready devices) .

If not Roon implement a _DSP Volume_ where they use DSP to change the volume, many users prefer not to use this (because you lose resolution - but also because the volume control won't match volume control if you play other sources through the same device). Again if using DSP volume - you don't need this PR.

Otherwise endpoints will have _Fixed Volume_ - in which case Roon won't display a volume slider at all - and you can't make volume changes.

So what this PR implements is useful in this last case and is an alternative to _DSP Volume_.

As well as controling an amplifier via an IR Blaster - I expect it will be useful to users playing to Airplay or Chrome endpoints, as well as speakers or amplifiers like Kef  LSX / LS50 where there is already an HA integration which provides volume control - but where it isn't integrated with roon.

It's not needed to control volume from HA, but the Roon browser is where most Roon users will choose their music (that's their USP) - so it's better to also be able to change the volume there.

MartinHjelmare commented:
I'm still not following. Won't there always be a volume level state attribute available in the Roon media player, updated from the Roon server side?

pavoni commented:
No - Roon itself shows this:-

<img width="533" alt="Screenshot 2023-09-06 at 15 28 26" src="https://github.com/home-assistant/core/assets/4487313/a76cd878-64b4-44df-a770-a05b916fcd4b">

And HA shows this (and won't let you change the volume - because Roon doesn't implement it)

<img width="567" alt="Screenshot 2023-09-06 at 15 31 33" src="https://github.com/home-assistant/core/assets/4487313/6f74c64b-e463-40c3-9dd1-135561f5dfdc">

pavoni commented:
BTW I can see that the HA roon integration could do with some better handling of some of the roon volume types.

I'll do that in an another PR, it's not connected with this - and will likely need a library upgrade.

pavoni commented:
Thanks @drewancil. I have been using it for a few days and if merged you're going to love it!

drewancil commented:
I downloaded the Roon integration code and merged this PR into my custom_components to test and am happy to report that its working very well for me.

My setup is a McIntosh amplifier with a Python app I wrote that  runs on a Pi and provides all McIntosh panel controls as well as state via RS232<->MQTT.  Controlling it from Roon is now possible with this PR which is great.  Does Roon provide any facilities to bring the volume state from HA into Roon?  I don't mean to muddy the communication in this PR, but wondered.  Im a little handy with Python, but Typescript is not one of my skills.  Thanks agian!

pavoni commented:
Generally volume controls mirror between a roon endpoint and the matching media player in roon.

However the volume control provided by this PR just implements +1  & -1 - and doesn’t not provide any tracking of overall level.

The roon api does provide some other options for external volume control where volume is tracked - so one of those may do what you want.

You might want to experiment with Pyroon directly…

Would likely need a follow on PR to implement (assuming this one is accepted)

pavoni commented:
Just wanted to check your current view @MartinHjelmare.
Hopefully with the support above you can see I'm not lost in a cul-de-sac on my own!

If you're OK in principal I will put the missing documentation PR together.

MartinHjelmare commented:
I'm checking with other core members before I come back here. :+1:

bdraco committed:
Fix implicit name in airzone_cloud (#101907)

joostlek committed:
Split Withings coordinators (#101766)

* Subscribe to Withings webhooks outside of coordinator

* Subscribe to Withings webhooks outside of coordinator

* Split Withings coordinator

* Split Withings coordinator

* Update homeassistant/components/withings/sensor.py

* Fix merge

* Rename MEASUREMENT_COORDINATOR

* Update homeassistant/components/withings/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Fix feedback

---------

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
Uncancel task when swallowing CancelledError (#101884)

bdraco committed:
Bump zeroconf to 0.116.0 (#101915)

bdraco committed:
Fix implicit device name in wiz switch (#101914)

bdraco committed:
Bump aioesphomeapi to 17.1.5 (#101916)

joostlek committed:
Improve Withings test fixtures (#101931)

goloveychuk committed:
Fix Setpoint in Matter climate platform (#101929)

fix matter max setpoint

bdraco committed:
Avoid conversion of timestamps in jwt auth (#101856)

cdce8p committed:
Revert aiohttp to 3.8.5 for Python 3.11 (#101932)

joostlek committed:
Add device info to Launch Library (#98767)

elmurato committed:
Add diagnostics support in Minecraft Server (#101787)

* Add diagnostics support

* Return diagnostics dict directly

* Use syrupy snapshots for assertions in diagnostics test

* Use parametrize for testing diagnostics

* Remove unnecessary side_effect in patch

raman325 committed:
Make Basic CC Z-Wave values a light (#101438)

joostlek committed:
Migrate Panasonic Viera to has entity name (#96746)

Co-authored-by: Franck Nijhof <git@frenck.dev>

ludeeus committed:
Update hass-nabucasa from 0.71.0 to 0.73.0 (#101939)

bachya committed:
Add more specific typing to OpenUV coordinator (#101952)

elmurato committed:
Set category and enabled by default of Minecraft Server sensors (#101943)

* Use set instead of list for supported server types in sensor platform

* Set sensor categories and enabled by default

* Set edition and version as diagnostic sensors

cdce8p committed:
Update pre-commit to 3.5.0 (#101956)

CFenner committed:
Fix types in ViCare integration (#101926)

* fix type

* fix return type

* fix type

* fix type

* Apply suggestions from code review

* Update climate.py

* Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update __init__.py

* Update climate.py

* Update __init__.py

* Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update binary_sensor.py

* Update button.py

* Update sensor.py

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

raman325 committed:
Update zwave issue repair strings (#101940)

bdraco committed:
Add some more typing to HomeKit (#101959)

bachya committed:
Simplify state update logic for OpenUV sensors (#101972)

* Clean up OpenUV entity state logic

* Reduce

* Remove old file

* Simplify

bdraco committed:
Avoid polling state machine for available state in HomeKit (#101799)

bdraco committed:
Update HomeKit humidifiers to handle current humidity (#101964)

bdraco committed:
Small cleanups to HomeKit thermostats (#101962)

bdraco committed:
Enable strict typing in HomeKit (#101968)

bdraco committed:
Bump aioesphomeapi to 17.2.0 (#101981)

* Bump aioesphomeapi to 17.2.0

changelog: https://github.com/esphome/aioesphomeapi/compare/v17.1.5...v17.2.0

* fix import from wrong module

joostlek committed:
Use aiowithings (#101819)

* Subscribe to Withings webhooks outside of coordinator

* Subscribe to Withings webhooks outside of coordinator

* Split Withings coordinator

* Split Withings coordinator

* Use aiowithings

* Use aiowithings

* Use aiowithings

* Update homeassistant/components/withings/sensor.py

* Merge

* Remove startdate

* Minor fixes

* Bump to 0.4.1

* Fix snapshot

* Fix datapoint

* Bump to 0.4.2

* Bump to 0.4.3

* Bump to 0.4.4

cgtobi committed:
Add service set_preset_mode_with_end_datetime in Netatmo integration (#101674)

* Add Netatmo climate service set_preset_mode_with_end_datetime to allow you to specify an end datetime for a preset mode

* Make end date optional

* address review comments

* Update strings

* Update homeassistant/components/netatmo/const.py

* Update homeassistant/components/netatmo/strings.json

---------

Co-authored-by: Adrien JOLY <adrien.joly-veillith@live.fr>
Co-authored-by: G Johansson <goran.johansson@shiftit.se>

joostlek committed:
Update Withings measurements incrementally after the first update (#102002)

* Update incrementally after the first update

* Update incrementally after the first update

bachya committed:
Add state translations for OpenUV UV Level sensor (#101978)

bdraco committed:
Improve performance of config/entity_registry/get* calls (#101984)

bdraco committed:
Cache emulated_hue local ip check (#102020)

cdce8p committed:
Update numpy to 1.26.1 (#102021)

joostlek committed:
Add suggested display precision to Withings (#102023)

bdraco committed:
Bump aioesphomeapi to 18.0.1 (#102028)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

joostlek committed:
Check for port in Withings webhook creation (#102026)

Check if port is really 443

joostlek committed:
Make Withings test sensors from entity registry (#102025)

Test entities from entity registry

bdraco committed:
Revert "Update numpy to 1.26.1" (#102036)

bdraco committed:
Bump zeroconf to 0.118.0 (#102015)

joostlek committed:
Clean up Withings webhook (#102038)

joostlek committed:
Dynamically add sensors for new measurements in Withings (#102022)

* Dynamically add sensors for new data points in Withings

* Dynamically add sensors for new data points in Withings

* Add test

* Change docstring

* Store new measurements

* Fix feedback

* Add test back

* Add test back

* Add test back

joostlek committed:
Mark Withings as cloud push (#102040)

gjohansson-ST committed:
Allow to remove devices in Sensibo (#101890)

bdraco committed:
Bump SQLAlchemy to 2.0.22 (#102033)

bdraco committed:
Bump HAP-python to 4.9.0 (#102055)

DeerMaximum committed:
Bump pynina to 0.3.3 (#101960)

joostlek committed:
Add Withings to strict-typing (#101761)

* Add Withings to strict-typing

* Rebase

natekspencer committed:
Adjust WeatherFlow air density sensor device class and unit (#101777)

wbyoung committed:
Fix date observed is not sent by AirNow (#101921) (#101977)

* Fix mixed up keys

These were accidentally the wrong values, but never passed on
to the end user.

* Fix date observed is not sent by AirNow (#101921)

allenporter committed:
Bump gcal_sync to 5.0.0 (#102010)

* Bump gcal_sync to 5.0.0

* Update snapshots

joostlek committed:
Add diagnostics to Withings (#102066)

joostlek committed:
Promote Withings to Platinum quality (#102069)

joostlek committed:
Add more measurement sensors to Withings (#102074)

bdraco committed:
Switch hassio to use iter_chunks (#102031)

allenporter committed:
Skip CalDAV calendars that do not support events (#102059)

dieselrabbit committed:
Bump screenlogicpy to v0.9.3 (#101957)

joostlek committed:
Address late Withings review (#102075)

gjohansson-ST committed:
Remove setup_platform for demo (#100867)

bdraco committed:
Move event permissions out of the websocket api into auth (#101975)

ollo69 committed:
Remove shorthand unique id in AsusWrt ScannerEntity (#102076)

Remove _attr_unique_id in AsusWrt ScannerEntity

bdraco committed:
Fix ingress sending an empty body for GET requests (#101917)

bdraco committed:
Reduce websocket event and state JSON construction overhead (#101974)

pnbruckner committed:
Fix google_maps same last_seen bug (#101971)

DCSBL committed:
Set Mac as connection to link HomeWizard devices on network (#101944)

bdraco committed:
Remove implict name check from Entity base class (#101905)

vangorra committed:
Remove code owner from withings (#102081)

* Remove code owner from withings

* Update CODEOWNERS

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

allenporter committed:
Fix bug in calendar state transitions (#102083)

abmantis committed:
Call disconnected callbacks from BT ESPHome client (#102084)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Use stdlib ip_address method in the network helper when compatible (#102019)

bdraco committed:
Bump aioesphomeapi to 18.0.3 (#102085)

bdraco committed:
Migrate ESPHome unique ids to new format (#99451)

gjohansson-ST committed:
Migrate unique id in Trafikverket Camera (#101937)

mkmer committed:
Bump Blinkpy to 0.22.2 in Blink (#98571)

CFenner committed:
Update device class, icons and names of Vicare binary sensors (#101476)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

cereal2nd committed:
Bump velbusaio to 2023.10.0 (#102100)

cereal2nd committed:
Correct sensor state attribute and device class in Velbus sensors (#102099)

rytilahti committed:
Allow model-specific lazy_discover setting for xiaomi_miio (#100490)

mkmer committed:
Address Blink late review (#102106)

* Address late review topics

* Missing await, optimize config_flow call

* Address proper mock for blink

* Address typing

TheJulianJES committed:
Bump zha-quirks to 0.0.105 (#102113)

elmurato committed:
Patch library instead of own code in Minecraft Server config flow tests (#102018)

* Patch library instead of own code in config flow tests

* Fix patch location

jpbede committed:
Make location types in co2signal translatable (#102127)

elupus committed:
Remove class argument in favor of class variables (zha) (#102117)

* Drop id_suffix class argument

* Use lowecase attribute

synesthesiam committed:
Add HassNevermind intent (bump intents package) (#102131)

* Add HassNevermind intent

* Bump intents package to 2023.10.16

blalor committed:
Add support for Govee H5055 (#100365)

Co-authored-by: J. Nick Koston <nick@koston.org>

jesserockz committed:
Send events for tts stream start/end (#102139)

bdraco committed:
Replace any expression in HomeKitWindowCover with a simple or (#102146)

bdraco committed:
Save previous unique id in entity registry when it changes (#102093)

bdraco committed:
Preserve HomeKit Accessory ID when entity unique id changes (#102123)

tronikos committed:
Bump opower to 0.0.36 (#102150)

emontnemery committed:
Fix menu in mysensors config flow (#102169)

emontnemery committed:
Fix menu in hassio repair flow (#102162)

elmurato committed:
Reach full init test coverage in Minecraft Server (#102013)

* Add full init test coverage

* Fix patch location

* Use contant for test unique ID

emontnemery committed:
Fix menu in homeassistant_hardware config flow (#102164)

emontnemery committed:
Fix menu in here_travel_time config flow (#102163)

emontnemery committed:
Fix menu in homeassistant_yellow config flow (#102166)

synesthesiam committed:
Don't warn about unknown pipeline events in ESPHome (#102174)

Don't warn about unknown events (debug)

dseven committed:
Explicitly set entity name for VenstarSensor (#102158)

VenstarSensor entity name

Set entity name using _attr_name instead of a name property, to
avoid warnings about name being implicitly set to None.

balloob committed:
Add some entity categories to Reolink (#102141)

elupus committed:
Add translation entiry support (zha) (#101909)

* zha: Add translation to binary sensors

* Add some small test for name

* Add translations for sensors

* Correct some tests

* Adjust summation key

* Add translation keys for button

* Add translation keys to climate

* Add translation keys for cover

* Add translation keys to fan

* Add translations to light

* Add translations for lock

* Add translation keys to number

* Add translationk keys to select

* Add translations for switch entities

* Add translation to alarm control panel

* Map to some more standard device classes

* Use shorter references

* Remove explicit name from identify button

* Correct tests

* Correction after rebase

TheJulianJES committed:
Remove unused variables in ZHA lighting cluster handler (#102138)

* Remove unused `UNSUPPORTED_ATTRIBUTE`

* Remove unused `CAPABILITIES_COLOR_TEMP`

* Use zigpy `ColorCapabilities` and remove `CAPABILITIES_COLOR_XY`

TheJulianJES committed:
Remove invalid attribute reporting for `enhanced_current_hue` in ZHA (#102137)

emontnemery committed:
Fix menu in knx config flow (#102168)

* Fix menu in knx config flow

* Update tests

* Fix strings.json

* Rename new menu steps for readabiltiy

---------

Co-authored-by: Matthias Alphart <farmio@alphart.net>

bdraco committed:
Only compute homekit_controller accessory_info when entity is added or config changes (#102145)

bdraco committed:
Cache construction of battery icon (#102194)

This was being built every time state was written. When
a robo vac is operating it writes state often which mean building
the icon string over and over again when it rarely changes.

bdraco committed:
Handle removal of accessories/services/chars in homekit_controller (#102179)

bdraco committed:
Bump aioesphomeapi to 18.0.6 (#102195)

bdraco committed:
Handle re-adding of accessories/services/chars in homekit_controller after removal (#102192)

chemelli74 committed:
Bump aiovodafone to 0.4.1 (#102180)

joostlek committed:
Bump aiowaqi to 2.1.0 (#102209)

MartinHjelmare commented:
We concluded that a device trigger or event entity is ok. An event entity is preferred today.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -27,6 +33,16 @@
     }
 )
 
+OPTIONS_SCHEMA = vol.Schema(
```
I don't think we need an options schema.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -136,6 +136,7 @@ async def async_added_to_hass(self) -> None:
             )
         )
         self._server.add_player_id(self.entity_id, self.name)
+        self._server.add_player_volume_hook(self.entity_id, self.name)
```
Can we create an event entity instead?

https://developers.home-assistant.io/docs/core/entity/event

MartinHjelmare commented on homeassistant/components/roon/device_trigger.py:
```diff
@@ -0,0 +1,83 @@
+"""Provides device triggers for RoonLabs music player."""
```
We don't need this if we add an event entity.

cdce8p committed:
Update mypy to 1.6.1 (#102210)

dupondje committed:
Fix DSMR max current device class (#102219)

Max current is CURRENT Device Class

The max current is shown as Amps. So need to change the device class to
CURRENT instead of POWER.

joostlek committed:
Change config entry title to Picnic (#102221)

joostlek committed:
Remove instances of title case in common strings (#102212)

MadMonkey87 committed:
Add sensors for myStrom plugs (#97024)

* support sensors for myStrom plugs

* added myStrom sensor to coveragerc

* some improvements from pr reviews

* adapt to the SensorEntityDescription pattern

* Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/mystrom/sensor.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

* Update homeassistant/components/mystrom/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update __init__.py

* Update const.py

* Update sensor.py

* Apply suggestions from code review

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: G Johansson <goran.johansson@shiftit.se>

janiversen committed:
Correct process_raw_value for modbus sensor (#102032)

frenck committed:
Update home-assistant/wheels to 2023.10.5 (#102243)

CFenner committed:
Get diagnostics of all devices in ViCare account (#102218)

* get diagnostics of all devices

* correct test data

* correct test data

* correct test data

dupondje committed:
Bump dsmr_parser to 1.3.0 (#102225)

bump dsmr_parser==1.3.0

frenck committed:
Update base image to 2023.10.0 (#102126)

bdraco committed:
Bump dbus-fast to 2.12.0 (#102206)

bdraco committed:
Bump zeroconf to 0.119.0 (#102207)

dependabot[bot] committed:
Bump actions/checkout from 4.1.0 to 4.1.1 (#102248)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

chemelli74 committed:
Add switch platform to Comelit SmartHome (#102233)

* Add switch platform to Comelit SmartHome

* add device class only when needed

* apply review comment

* small cleanup for light platform

* update functions description

* fix list of values

bdraco committed:
Bump bluetooth-data-tools to 1.13.0 (#102208)

stefanroelofs committed:
Remove Shiftr integration (#102224)

cereal2nd committed:
Bump velbusaio to 2023.10.1 (#102178)

abmantis committed:
Add better connection management for Idasen Desk (#102135)

piitaya committed:
Return 'None' for light attributes when off instead of removing them (#101946)

bdraco committed:
Bump home-assistant-bluetooth to 1.10.4 (#102268)

bdraco committed:
Bump orjson to 3.9.9 (#102267)

tronikos committed:
Bump opower to 0.0.37 (#102265)

Lash-L committed:
Bump Python-Roborock to 0.35.0 (#102275)

frenck committed:
Add CodeQL CI Job (#102273)

dupondje committed:
Remove unused dsmr sensors (#102223)

bdraco committed:
Make group _update_at_start a callback (#102286)

bdraco committed:
Refactor automation trigger attachment to avoid creating a closure (#102288)

Noltari committed:
Handle timeouts on AEMET init (#102289)

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

bdraco committed:
Handle ATTR_HS_COLOR being None in HomeKit (#102290)

CFenner committed:
Reuse function to check feature support on ViCare devices (#102211)

* check feature support in utils function

* rename parameters

* restore severity

* reorder parameters

* Update .coveragerc

joostlek committed:
Patch platform in Withings sensor test (#102155)

* Patch platform in Withings sensor test

* Fix feedback

jbouwh committed:
Do not fail mqtt entry on single platform config schema error (#101373)

* Do not fail mqtt entry on platform config

* Raise on reload with invalid config

* Do not store issues

* Follow up

joostlek committed:
Move WAQI state attributes to separate sensors (#101217)

* Migrate WAQI to has entity name

* Split WAQI extra state attributes into separate sensors

* Split WAQI extra state attributes into separate sensors

* Fix test

* Support new aiowaqi

* Bump aiowaqi to 2.1.0

* Add nephelometry as possible value

* Fix test

emontnemery committed:
Validate steps in Flowhandler (#102152)

* Validate steps in Flowhandler

* Move validation to FlowManager._async_handle_step

* Fix _raise_if_not_has_step

* Fix config_entries tests

* Fix tests

* Rename

* Add test

emontnemery committed:
Rename `gather_with_concurrency` to `gather_with_limited_concurrency` (#102241)

* Rename gather_with_concurrency to gather_with_limited_concurrency

* Update test

Xitee1 committed:
Add more diagnostic sensors to iRobot (#84995)

Co-authored-by: 930913 <3722064+930913@users.noreply.github.com>
Co-authored-by: Xitee <59659167+Xitee1@users.noreply.github.com>
Co-authored-by: J. Nick Koston <nick@koston.org>
Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

eifinger committed:
Use snapshots in weather tests (#102297)

elupus committed:
Standardize zha attribute member name (#102182)

* Correct missed translation

* Standardize on _attribute for zha

jbouwh committed:
Do not fail MQTT setup if lights configured via yaml can't be validated (#101649)

* Add light

* Deduplicate code

* Follow up comment

jbouwh committed:
Do not fail MQTT setup if binary sensors configured via yaml can't be validated (#102300)

Add binary_sensor

jbouwh committed:
Do not fail MQTT setup if camera's configured via yaml can't be validated (#102302)

Add camera

jbouwh committed:
Do not fail MQTT setup if device trackers configured via yaml can't be validated (#102308)

Add device_tracker

edenhaus committed:
Fix KeyError in derivative and integration (#102294)

GeoffAtHome committed:
Smart plugs appear as Switches and Binary Sensors (#102112)

jbouwh committed:
Do not fail MQTT setup if scenes configured via yaml can't be validated (#102317)

Add scene

bdraco committed:
Reduce internal property lookups needed to write number entity state (#102281)

jbouwh committed:
Do not fail MQTT setup if fans configured via yaml can't be validated (#102310)

Add fan

jbouwh committed:
Do not fail MQTT setup if images configured via yaml can't be validated (#102313)

Add image

jbouwh committed:
Do not fail MQTT setup if sirens configured via yaml can't be validated (#102319)

Add siren

jbouwh committed:
Do not fail MQTT setup if text's configured via yaml can't be validated (#102322)

Add text

chemelli74 committed:
Bump aiocomelit to 0.3.0 (#102340)

* Bump aiocomelit to 0.3.0

* missing string

jbouwh committed:
Fix suggested UOM cannot be set for dsmr entities (#102134)

* Supply dsmr entities jit on first telegram

* Stale docstr

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Simplify tuple type

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

frenck committed:
Don't run CodeQL on PRs (#102342)

jbouwh committed:
Do not fail MQTT setup if vacuum's configured via yaml can't be validated (#102325)

Add vacuum

bdraco committed:
Remove update_before_add from roomba (#102337)

jbouwh committed:
Do not fail MQTT setup if climate's configured via yaml can't be validated (#102303)

Add climate

joostlek committed:
Use entity descriptions in Roomba (#102323)

Co-authored-by: J. Nick Koston <nick@koston.org>

thatso committed:
Improve wording in NUT (#102353)

bdraco committed:
Improve typing in roomba integration (#102350)

joostlek committed:
Import Comelit state from library (#102356)

jbouwh committed:
Do not fail MQTT setup if select's configured via yaml can't be validated (#102318)

Add select

jbouwh committed:
Do not fail MQTT setup if switches configured via yaml can't be validated (#102320)

Add switch

bdraco committed:
Bump aiohomekit to 3.0.6 (#102341)

jbouwh committed:
Do not fail MQTT setup if humidifiers configured via yaml can't be validated (#102312)

Add humidifier

joostlek committed:
Address MyStrom late review (#102306)

* Address MyStrom late review

* Address MyStrom late review

jbouwh committed:
Do not fail MQTT setup if locks configured via yaml can't be validated (#102315)

Add lock

frenck committed:
Remove dev API docs from repository (#102274)

jbouwh committed:
Do not fail MQTT setup if lawn mowers configured via yaml can't be validated (#102314)

puddly committed:
Bump ZHA dependencies (#102358)

bdraco committed:
Use new lookup methods for homekit_controller (#102278)

edenhaus committed:
Fix multilevel reference translations (#102338)

jbouwh committed:
Do not fail MQTT setup if buttons configured via yaml can't be validated (#102301)

cereal2nd committed:
Bump pyduotecno to 2023.10.1 (#102344)

jbouwh committed:
Do not fail MQTT setup if water heaters configured via yaml can't be validated (#102326)

TheJulianJES committed:
Fix ZHA `power_factor` attribute not initialized (#102133)

puddly committed:
Clean up stale ZHA database listener when reconnecting to radio (#101850)

Kane610 committed:
Fix UniFi client tracker entities being unavailable when away on restart (#102125)

bdraco committed:
Reduce overhead to write HomeKit Controller state (#102365)

jbouwh committed:
Do not fail MQTT setup if update entities configured via yaml can't be validated (#102324)

jbouwh committed:
Do not fail MQTT setup if numbers configured via yaml can't be validated (#102316)

Add number

jbouwh committed:
Do not fail MQTT setup if covers configured via yaml can't be validated (#102304)

Add cover

frenck committed:
Bump toonapi to 0.3.0 (#102369)

c0ffeeca7 committed:
Rename Twitter to X (#102214)

Noltari committed:
Implement Airzone Cloud Installation climate support (#101090)

Co-authored-by: J. Nick Koston <nick@koston.org>

jbouwh committed:
Do not fail MQTT setup if events or sensors configured via yaml can't be validated (#102309)

* Add event and sensor

* Cleanup unused code

* Schema cannot be None for supported platform

Noltari committed:
Improve Airzone Cloud tests (#102377)

jbouwh committed:
Rename mqtt entry setup helpers to reflect their purpose (#102378)

Rename mqtt entry setup helpers

Archomeda committed:
Fix Spotify media position update value (#100044)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jbouwh committed:
Improve test creating mqtt certificate files (#102380)

* Improve test creating mqtt certificate files

* Split tests

* Cleanup and de-duplicate

* Update tests/components/mqtt/test_util.py

jpbede committed:
Bump pydiscovergy to 2.0.5 (#102354)

joostlek committed:
Bump vehicle to 2.0.0 (#102379)

Noltari committed:
Update aioairzone-cloud to v0.2.5 (#102382)

Noltari committed:
Update aioairzone to v0.6.9 (#102383)

eifinger committed:
Use snapshots in calendar tests (#102299)

Sjorsa committed:
Fix typo in fastdotcom strings (#102384)

jbouwh committed:
Fix error handling on subscribe when mqtt is not initialized (#101832)

Marco4223 committed:
Get all playlist items from sonos devices (#100924)

dupondje committed:
Add peak usage sensors to dsmr (#102227)

cereal2nd committed:
Create a binary sensor for each Duotecno virtual unit (#102347)

chatziko committed:
Use action response in intent_script speech template (#96256)

bdraco committed:
Bump ulid-transform to 0.9.0 (#102272)

bdraco committed:
Bump fnv-hash-fast to 0.5.0 (#102271)

kpine committed:
Fix temperature setting for multi-setpoint z-wave device (#102395)

* Fix temperature setting for multi-setpoint z-wave device

* Add missing fixture file

* Apply suggestions from code review

---------

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Bump aioesphomeapi to 18.0.7 (#102399)

bdraco committed:
Reduce number of test states in big purge test to fix CI (#102401)

frenck committed:
Update pvo to 2.0.0 (#102398)

frenck committed:
Update guppy3 to 3.1.4 (#102400)

jbouwh committed:
Refactor mqtt entity cleanup on reload (#102375)

cgarwood committed:
Add Enphase charge from grid switch (#102394)

* Add Enphase charge from grid switch

* review comments

frenck committed:
Update elgato to 5.0.0 (#102405)

Noltari committed:
Update aioairzone-cloud to v0.2.7 (#102406)

frenck committed:
Bump twentemilieu to 2.0.0 (#102407)

frenck committed:
Update wled to 0.17.0 (#102413)

frenck committed:
Update psutil to 5.9.6 (#102416)

bdraco committed:
Bump aiohomekit to 3.0.7 (#102408)

frenck committed:
Update apprise to 1.6.0 (#102417)

jpbede committed:
Use dataclass to carry data in ping (#99803)

Co-authored-by: J. Nick Koston <nick@koston.org>

frenck committed:
Update Pillow to 10.1.0 (#102419)

frenck committed:
Update ruff to v0.1.1 (#102421)

bdraco committed:
Only callback when value or status changes for processing HKC events (#102370)

raman325 committed:
Add support for zwave_js event entities (#102285)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

cgarwood committed:
Add support for changing Enphase battery backup modes (#102392)

rikroe committed:
Bump bimmer_connected to 0.14.2 (#102426)

Co-authored-by: rikroe <rikroe@users.noreply.github.com>

frenck committed:
Update black to 23.10.0 (#102420)

bdraco committed:
Bump pyenphase to 1.13.1 (#102431)

balloob committed:
Set Reolink record switch as config (#102439)

Reolink: record switch as config

pavoni committed:
Refactor to use Event entities.

pavoni committed:
Add tests.

pavoni committed:
Merge branch 'roon_volume' of https://github.com/home-assistant/core into roon_volume

pavoni commented:
I messed up a rebase - so have opened a replacemeny PR https://github.com/home-assistant/core/pull/102470
